File: code/python/剑指offer07.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        def buildtree(preOrder,inOrder,root):
            if len(preOrder)==0:
                return None
            node = TreeNode(preOrder[0])
            i = 0
            while inOrder[i]!=preOrder[0]:
                i+=1
            node.left = buildtree(preOrder[1:i+1],inOrder[:i],node)
            node.right = buildtree(preOrder[i+1:],inOrder[i+1:],node)
            return node
        root = buildtree(preorder,inorder,None)
        return root

File: code/java/剑指offer07.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode buildNode(int[] preorder,int preL,int preR,int[] inorder,int inL,int inR){
        if(preL>preR||inL>inR) return null;
        TreeNode newNode = new TreeNode(preorder[preL]);
        int i = 0;
        while(inorder[inL+i]!=preorder[preL]) i+=1;
        newNode.left = buildNode(preorder,preL+1,preL+i,inorder,inL,inL+i-1);
        newNode.right = buildNode(preorder,preL+i+1,preR,inorder,inL+i+1,inR);

        return newNode;
    }

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        TreeNode root = buildNode(preorder,0,preorder.length-1,inorder,0,inorder.length-1);
        return root;
    }
}